### PR TITLE
Stub some missing AM Ticket functions

### PR DESCRIPTION
**Pull request by PabloMK7**
Stubs enough AM functionality so that loading the eShop works, and performing a system update returns an error instead of crashing.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7172)

### CODE REVIEW

File: src/core/hle/service/am/am.cpp [hunk 1]
```diff
@@ -374,6 +374,46 @@ bool CIAFile::Close() const {
...
+}
+
+ResultVal<std::size_t> TicketFile::Read(u64 offset, std::size_t length, u8* buffer) const {
+    UNIMPLEMENTED();
+    return length;
+}
+
+ResultVal<std::size_t> TicketFile::Write(u64 offset, std::size_t length, bool flush,
+                                         const u8* buffer) {
+
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:390]:
```suggestion
```

File: src/core/hle/service/am/am.cpp [hunk 2]
```diff
@@ -374,6 +374,46 @@ bool CIAFile::Close() const {
...
+    Close();
+}
+
+ResultVal<std::size_t> TicketFile::Read(u64 offset, std::size_t length, u8* buffer) const {
+    UNIMPLEMENTED();
+    return length;
+}
+
+ResultVal<std::size_t> TicketFile::Write(u64 offset, std::size_t length, bool flush,
+                                         const u8* buffer) {
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:389]: can this be a span?

File: src/core/hle/service/am/am.cpp [hunk 3]
```diff
@@ -374,6 +374,46 @@ bool CIAFile::Close() const {
...
+
+    written += length;
+    data.resize(written);
+    memcpy(data.data() + offset, buffer, length);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:393]:
```suggestion
    std::memcpy(data.data() + offset, buffer, length);
```

File: src/core/hle/service/am/am.cpp [hunk 4]
```diff
@@ -374,6 +374,46 @@ bool CIAFile::Close() const {
...
+}
+
+bool TicketFile::SetSize(u64 size) const {
+    return false;
+}
+
+bool TicketFile::Close() const {
+    FileSys::Ticket ticket;
+    if (ticket.Load(data, 0) == Loader::ResultStatus::Success) {
+        LOG_WARNING(Service_AM, "Discarding ticket for 0x{:016X}.", ticket.GetTitleID());
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:408]: using #x in the format string, you should be able to remove the manual `0x`

File: src/core/hle/service/am/am.cpp [hunk 5]
```diff
@@ -1177,6 +1221,16 @@ void Module::Interface::NeedsCleanup(Kernel::HLERequestContext& ctx) {
     rb.Push<bool>(false);
 }
 
+void Module::Interface::DoCleanup(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+    const auto media_type = rp.Pop<u8>();
+
+    LOG_DEBUG(Service_AM, "(STUBBED) media_type=0x{:02x}", media_type);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1228]:
```suggestion
    LOG_DEBUG(Service_AM, "(STUBBED) called, media_type=0x{:02x}", media_type);
```

File: src/core/hle/service/am/am.cpp [hunk 6]
```diff
@@ -1188,6 +1242,45 @@ void Module::Interface::QueryAvailableTitleDatabase(Kernel::HLERequestContext& c
...
+void Module::Interface::GetPersonalizedTicketInfoList(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+    [[maybe_unused]] u32 ticket_count = rp.Pop<u32>();
+    [[maybe_unused]] auto& buffer = rp.PopMappedBuffer();
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
+    rb.Push(RESULT_SUCCESS); // No error
+    rb.Push(0);
+
+    LOG_WARNING(Service_AM, "(STUBBED) ticket_count={}", ticket_count);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1254]:
```suggestion
    LOG_WARNING(Service_AM, "(STUBBED) called, ticket_count={}", ticket_count);
```

File: src/core/hle/service/am/am.cpp [hunk 7]
```diff
@@ -1188,6 +1242,45 @@ void Module::Interface::QueryAvailableTitleDatabase(Kernel::HLERequestContext& c
...
+void Module::Interface::GetNumImportTitleContextsFiltered(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+    u8 media_type = rp.Pop<u8>();
+    u8 filter = rp.Pop<u8>();
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
+    rb.Push(RESULT_SUCCESS); // No error
+    rb.Push(0);
+
+    LOG_WARNING(Service_AM, "(STUBBED) media_type={}, filter={}", media_type, filter);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1266]:
```suggestion
    LOG_WARNING(Service_AM, "(STUBBED) called, media_type={}, filter={}", media_type, filter);
```

File: src/core/hle/service/am/am.cpp [hunk 8]
```diff
@@ -1188,6 +1242,45 @@ void Module::Interface::QueryAvailableTitleDatabase(Kernel::HLERequestContext& c
...
+    u8 media_type = rp.Pop<u8>();
+    u8 filter = rp.Pop<u8>();
+    auto& buffer = rp.PopMappedBuffer();
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 2);
+    rb.Push(RESULT_SUCCESS); // No error
+    rb.Push(0);
+    rb.PushMappedBuffer(buffer);
+
+    LOG_WARNING(Service_AM, "(STUBBED) media_type={}, filter={}", media_type, filter);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1281]:
```suggestion
    LOG_WARNING(Service_AM, "(STUBBED) called, media_type={}, filter={}", media_type, filter);
```

File: src/core/hle/service/am/am.cpp [hunk 9]
```diff
@@ -1188,6 +1242,45 @@ void Module::Interface::QueryAvailableTitleDatabase(Kernel::HLERequestContext& c
...
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
+    rb.Push(RESULT_SUCCESS); // No error
+    rb.Push(0);
+
+    LOG_WARNING(Service_AM, "(STUBBED) media_type={}, filter={}", media_type, filter);
+}
+
+void Module::Interface::GetImportTitleContextListFiltered(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+    [[maybe_unused]] u32 count = rp.Pop<u32>();
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1271]: these can all be const

File: src/core/hle/service/am/am.cpp [hunk 10]
```diff
@@ -1674,6 +1767,31 @@ void Module::Interface::GetMetaDataFromCia(Kernel::HLERequestContext& ctx) {
     rb.PushMappedBuffer(output_buffer);
 }
 
+void Module::Interface::BeginImportTicket(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+
+    // Create our TicketFile handle for the app to write to
+    const FileSys::Path path = {};
+    auto file =
+        std::make_shared<Service::FS::File>(am->kernel, std::make_unique<TicketFile>(), path);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp:1776]: shouldn't this work as well?
```suggestion
        std::make_shared<Service::FS::File>(am->kernel, std::make_unique<TicketFile>(), {}});
```

**Review by FearlessTobi — CHANGES_REQUESTED** [review]

File: src/core/hle/service/am/am.cpp (hunk 2 above)
Comment (PabloMK7) [reply src/core/hle/service/am/am.cpp:389]: Nope, this is an overriden method from the FileBackend class, would need to change that class method definition as well as all the children classes, which is out of the scope of this PR.

**Review by Steveice10 — APPROVED** [review]

**Pull request merged**